go-pkg/ds/permission.go

```diff
@@ -22,6 +28,62 @@ func (u *User) addKey(blob *protods.DataBlob, newDataOwner *User) *protods.DataB
 // 1. Generate an AES-key
 // 2. Add an entry to "Secrets"
 // 3. Encrypt data with AES key
-func (u *User) createDataBlob(data []byte) *protods.DataBlob {
-	panic("not implemented")
+func (u *User) createDataBlob(data []byte) (*protods.DataBlob, error) {
+	hiddenDataKey, AESKey, err := u.generateHiddenKey()
+	if err != nil {
+		return nil, err
+	}
+	dataLen := len(data)
+	iv := encrypt.GenCBCIv()
+
+	encryptedData, err := encrypt.Encrypt(AESKey, iv, data)
+	if err != nil {
+		return nil, errors.New("encryption of data with data key failed")
+	}
+
+	encryptedDataHash := ethereum.Keccak256Hash(encryptedData)
+
+	dataBlob := protods.DataBlob{
+		DataLen:           uint64(dataLen),
+		Iv:                iv,
+		EncryptedDataHash: encryptedDataHash[:],
+		EncryptedData:     encryptedData,
+		Keys:              []*protods.HiddenDataKey{hiddenDataKey},
+	}
+
+	return &dataBlob, nil
+}
+
+func (u *User) generateHiddenKey() (*protods.HiddenDataKey, []byte, error) {
+	//HiddenDataKey creation
+	dataAESKey := encrypt.GenAes128Key()
+
+	ephemeralPrivKey, err := ethereum.GenerateKey()
+	if err != nil {
+		return nil, nil, errors.New("creation of ephemeral private key failed")
+	}
+	sharedSecretX, sharedSecretY := ephemeralPrivKey.PublicKey.ScalarMult(
+		ephemeralPrivKey.PublicKey.X,
+		ephemeralPrivKey.PublicKey.Y,
+		u.privkey.D.Bytes(),
+	)
+
+	sharedSecret := elliptic.Marshal(ethereum.S256(), sharedSecretX, sharedSecretY)
```

Comment:
Let's keep this simpler and without ambiguity on the `ethereum.S256()` byte representation.

1. Get the the first coordinate of the the secret point (x). This coordinate is a 32 byte integer.
2. sha256(point_x in byte form)
3. Take the the first 128 bits as `ephemeralAesKey`.